pkgs/tools/security/sha1collisiondetection/default.nix

```diff
@@ -0,0 +1,39 @@
+{ stdenv, fetchFromGitHub, libtool }:
+
+stdenv.mkDerivation  rec {
+  pname = "sha1collisiondetection";
+  version = "2017-02-21";
+  name = "${pname}-unstable-${version}";
+
+  src = fetchFromGitHub {
+    owner = "cr-marcstevens";
+    repo = "${pname}";
+    rev = "40ccf5e3537ff6b2e5ceac5747f376a2ef430bec";
+    sha256 = "1k2qjarcx88ndc4yl8xnr1l3j74w1qacvyrhv4ms9hndy8lrsm0l";
+  };
+
+  patchPhase = ''substituteInPlace Makefile --replace "shell arch" "uname -m"'';
+
+  buildInputs = [ libtool ];
+
+  installPhase = ''
+    mkdir -p $out/bin $out/lib
+    cp bin/sha1dcsum bin/sha1dcsum_partialcoll $out/bin
+    cp bin/libdetectcoll.la $out/lib/
+    '';
+
+  meta = with stdenv.lib; {
+    description = "Library and command line tool to detect SHA-1 collision";
+    longDescription = ''
+      This library and command line tool were designed as near drop-in
+      replacements for common SHA-1 libraries and sha1sum. They will
+      compute the SHA-1 hash of any given file and additionally will
+      detect cryptanalytic collision attacks against SHA-1 present in
+      each file. It is very fast and takes less than twice the amount
```

Comment:
@edolstra: The description is from the author (Marc Stevens from CWI) himself. The cool trick the tools perform is counter-cryptanalytics. Imagine some tarball with source code which is crafted specifically to perform an attack, and inject through a man in the middle attack. How nix plans to deal with the source file once it has passed the test (say: configure, make, make install) is fully transparent, and so it will not be difficult to  allow attacker controlled code to end up in a running system. Manipulating the sha1sum value of the fake source object is after all possible (as per the paper) and would result in its sha1sum to be the same as the original object (a "cryptanalytic collision attack"). However, and fortunate for us, that procedure leaves traces. With these tools those traces show up like on an x-ray: _each file_ is checked for _cryptanalytic collision attacks_ after the regular hash is computed. But instead of making it a separate tool, this does something even smarter: it replaces the original tool with a 'secure' version. So it does exactly what the original sha1sum does, but then proceeds to a second stage checking the countercryptanalytic indicators. If an attack (i.e. a manipulated sha1 hash) is found, rather than crying foul it will just return an alternative hash - making the attacked process reject the manipulated file.